### PR TITLE
Feature/features order

### LINKED ISSUE #64: Add 'order' field to project features
A Chingu member needs to have the ability to order features within each 'bucket' (Must Have, Should Have, Nice to Have).

See Trello card for more details.

### PR DESCRIPTION
**Pull request by smurph7894**
# Description

Updated Feature Module to include ordering per team and category.  Note all but one service has changes to take into account their order. UpdateFeatureOrderAndCategory is open to team members to manipulate per FE request. 

## Issue link

Fixes #64 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Tested in swagger.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by cherylli — COMMENTED** [review]
with the new changes we would have a PATCH endpoint for updating "description" and another endpoint for updating the order, am I correct in understanding this? 

Would it be better to have 1 endpoint where you can change both? depends on what is in the request body

**Comment by smurph7894** [comment]
> with the new changes we would have a PATCH endpoint for updating "description" and another endpoint for updating the order, am I correct in understanding this?
> 
> Would it be better to have 1 endpoint where you can change both? depends on what is in the request body

I talked to  Jane to get requirements and there are 2 separate updates needed. Our normal one for the creator to update the description and another for all team members to be able to drag and drop features to different orders and categories.

**Comment by cherylli** [comment]
> > with the new changes we would have a PATCH endpoint for updating "description" and another endpoint for updating the order, am I correct in understanding this?
> > Would it be better to have 1 endpoint where you can change both? depends on what is in the request body
> 
> I talked to Jane to get requirements and there are 2 separate updates needed. Our normal one for the creator to update the description and another for all team members to be able to drag and drop features to different orders and categories.

you would still be able to do both in the same endpoint. If we have to add more things to update, would we make a new endpoint for each thing we need to update? What does everyone else think?

**Comment by smurph7894** [comment]
> > > with the new changes we would have a PATCH endpoint for updating "description" and another endpoint for updating the order, am I correct in understanding this?
> > > Would it be better to have 1 endpoint where you can change both? depends on what is in the request body
> > 
> > 
> > I talked to Jane to get requirements and there are 2 separate updates needed. Our normal one for the creator to update the description and another for all team members to be able to drag and drop features to different orders and categories.
> 
> you would still be able to do both in the same endpoint. If we have to add more things to update, would we make a new endpoint for each thing we need to update? What does everyone else think?

It would depend on what the update is and what makes sense for the situation if you add another endpoint or not. 

I have them currently as two separate as it made sense since the UI actions are separate. Also the order and category update is more complex and both would be easier to maintain if kept to single responsibility endpoints. I felt more conditionals being added to combine them would add more complexity and room for error. Is there a benefit to combining?

**Comment by cherylli** [comment]
> > > > with the new changes we would have a PATCH endpoint for updating "description" and another endpoint for updating the order, am I correct in understanding this?
> > > > Would it be better to have 1 endpoint where you can change both? depends on what is in the request body
> > > 
> > > 
> > > I talked to Jane to get requirements and there are 2 separate updates needed. Our normal one for the creator to update the description and another for all team members to be able to drag and drop features to different orders and categories.
> > 
> > 
> > you would still be able to do both in the same endpoint. If we have to add more things to update, would we make a new endpoint for each thing we need to update? What does everyone else think?
> 
> It would depend on what the update is and what makes sense for the situation if you add another endpoint or not.
> 
> I have them currently as two separate as it made sense since the UI actions are separate. Also the order and category update is more complex and both would be easier to maintain if kept to single responsibility endpoints. I felt more conditionals being added to combine them would add more complexity and room for error. Is there a benefit to combining?

no, just feel cleaning to have one endpoint doing all the updating. At first glance, /reorder feels like it should just do re-ordering but it also updates the featureCategory. 

or I felt it was not clear why there are 2 endpoints to update different things

File: src/features/dto/update-feature-order-and-category.dto.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { IsNotEmpty, IsOptional } from "class-validator";
+import { ApiProperty, ApiPropertyOptional } from "@nestjs/swagger";
+
+export class UpdateFeatureOrderAndCategoryDto {
+    @IsNotEmpty()
+    @ApiProperty()
+    order: number;
```
Comment (cherylli) [inline src/features/dto/update-feature-order-and-category.dto.ts:7]: I think we need to add @IsNumber() here

**Review by cherylli — APPROVED** [review]
Works fine but no error message when an invalid categoryId is entered

**Comment by cherylli** [comment]
[CGD-64: Add "order" field to project features](https://trello.com/c/mQeNfmbO/71-cgd-64-add-order-field-to-project-features)

**Comment by kennytram** [comment]
> endpoint for each thing we need to update? What does everyone else think?

The first PATCH endpoint only lets the creator to update the feature, whereas this PATCH endpoint will let anyone on the team to update the order.

I might be wrong because I haven't fully read Sarah's changes, but that's what I did on my end.

**Review by curtwl — DISMISSED** [review]
lgtm, reordering worked as expected

**Review by curtwl — APPROVED** [review]
LGTM. I tried some edge cases like ties and everything looked good.

**Review by cherylli — APPROVED** [review]
works fine for me locally

**Pull request merged**